File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/methods/userInteraction/PinsChangeMethodHandler.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.methods.userInteraction

import ch.nevis.mobile.authentication.sdk.react.MethodChannelHandler
import ch.nevis.mobile.authentication.sdk.react.cache.OperationCache
import ch.nevis.mobile.authentication.sdk.react.cache.operation.PinChangeOperation
import ch.nevis.mobile.authentication.sdk.react.cache.state.PinChange
import ch.nevis.mobile.authentication.sdk.react.methods.MethodHandler
import ch.nevis.mobile.authentication.sdk.react.model.ReactMethodName
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.ChannelInMessage
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.PinsChangeMessage
import ch.nevis.mobile.sdk.api.MobileAuthenticationClient

class PinsChangeMethodHandler : MethodHandler() {
    override fun execute(
        client: MobileAuthenticationClient,
        channelInMessage: ChannelInMessage,
    ) {
        val message = validate<PinsChangeMessage>(channelInMessage)
        val operation = OperationCache.getOperation<PinChangeOperation>(message.operationId)
        val state = operation.state as? PinChange
            ?: throw IllegalStateException("Operation state is missing or not a Pin change state!")
        val oldPin = message.oldPin?.toCharArray()
        val newPin = message.newPin?.toCharArray()
        state.pinChangeHandler?.pins(oldPin, newPin)
        MethodChannelHandler.resolve(ReactMethodName.PINS_CHANGE, null)
    }
}

File: ios/Methods/UserInteraction/PinsChangeMethodHandler.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

struct PinsChangeMethodHandler: MethodHandler {
	func execute(using _: MobileAuthenticationClient, with message: ChannelInMessage) throws {
		let message: PinsChangeMessage = try validate(message: message)
		let operation: PinChangeOperation = try OperationCache.shared.read(by: message.operationId)
		guard let state = operation.state as? PinChange else {
			throw PluginError.illegalState("Pin change state is expected!")
		}

		state.handler?.pins(message.oldPin, message.newPin)
		
		MethodChannelHandler.shared.resolve(method: .pinsChange, message: nil)
	}
}
